From: Rene Kita <mail@rkta.de>
Subject: Enable folding in pre tags

Some bug trackers are really annoying to read, because they place
comments in <pre> elements. They usually provide some CSS to make it
readable.

As we do not talk CSS, add a command to reload the page folding those
long lines. This is a one-time operation. Additionally add an option to
make this behaviour permanent.
---
 doc-de/README.func |    1 +
 doc/README.func    |    1 +
 file.c             |    7 +++++--
 fm.h               |    1 +
 main.c             |   11 +++++++++++
 proto.h            |    1 +
 rc.c               |    3 +++
 7 files changed, 23 insertions(+), 2 deletions(-)

--- a/file.c
+++ b/file.c
@@ -30,6 +30,7 @@
 
 #define MAX_INPUT_SIZE 80 /* TODO - max should be screen line length */
 
+extern int fold_pre;
 static int frame_source = 0;
 static int need_number = 0;
 
@@ -2567,7 +2568,7 @@ check_breakpoint(struct readbuffer *obuf
     int tlen, len = obuf->line->length;
 
     append_tags(obuf);
-    if (pre_mode)
+    if (pre_mode && !fold_pre)
 	return;
     tlen = obuf->line->length - len;
     if (tlen > 0
@@ -6586,7 +6587,8 @@ HTMLlineproc0(char *line, struct html_fe
 			proc_mchar(obuf, 1, delta, &str, mode);
 		}
 		if (obuf->flag & (RB_SPECIAL & ~RB_PRE_INT))
-		    continue;
+		    if (!fold_pre)
+			continue;
 	    }
 	    else {
 		if (!IS_SPACE(*str))
@@ -6801,6 +6803,7 @@ loadHTMLBuffer(URLFile *f, Buffer *newBu
 	if (src)
 	    newBuf->sourcefile = tmp->ptr;
     }
+    fold_pre |= FoldPre;
 
     loadHTMLstream(f, newBuf, src, newBuf->bufferprop & BP_FRAME);
 
--- a/fm.h
+++ b/fm.h
@@ -1044,6 +1044,7 @@ global int ignore_null_img_alt init(TRUE
 #define DISPLAY_INS_DEL_FONTIFY	2
 global int displayInsDel init(DISPLAY_INS_DEL_NORMAL);
 global int FoldTextarea init(FALSE);
+global int FoldPre init(FALSE);
 global int FoldLine init(FALSE);
 #define DEFAULT_URL_EMPTY	0
 #define DEFAULT_URL_CURRENT	1
--- a/main.c
+++ b/main.c
@@ -128,6 +128,7 @@ static int searchKeyNum(void);
 
 int enable_inline_image;
 extern int opt_cols;
+int fold_pre;
 
 static void
 fversion(FILE * f)
@@ -4895,6 +4896,14 @@ DEFUN(vwSrc, SOURCE VIEW, "Toggle betwee
     displayBuffer(Currentbuf, B_NORMAL);
 }
 
+DEFUN(foldPre, FOLD_PRE, "Fold long lines in <pre> elements")
+{
+    fold_pre = 1;
+    Currentbuf->need_reshape = TRUE;
+    displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    fold_pre = FoldPre;
+}
+
 /* reload */
 DEFUN(reload, RELOAD, "Load current document anew")
 {
@@ -4922,6 +4931,7 @@ DEFUN(reload, RELOAD, "Load current docu
 	disp_err_message("Can't reload stdin", TRUE);
 	return;
     }
+    fold_pre = FoldPre;
     copyBuffer(&sbuf, Currentbuf);
     if (Currentbuf->bufferprop & BP_FRAME &&
 	(fbuf = Currentbuf->linkBuffer[LB_N_FRAME])) {
@@ -5018,6 +5028,7 @@ DEFUN(reload, RELOAD, "Load current docu
 	restorePosition(Currentbuf, &sbuf);
     }
     displayBuffer(Currentbuf, B_FORCE_REDRAW);
+    fold_pre = FoldPre;
 }
 
 /* reshape */
--- a/proto.h
+++ b/proto.h
@@ -98,6 +98,7 @@ extern void peekURL(void);
 extern void peekIMG(void);
 extern void curURL(void);
 extern void vwSrc(void);
+extern void foldPre(void);
 extern void reload(void);
 extern void reshape(void);
 extern void chkURL(void);
--- a/rc.c
+++ b/rc.c
@@ -98,6 +98,7 @@ static int OptionEncode = FALSE;
 #define CMT_DISP_BORDERS N_("Display table borders, ignore value of BORDER")
 #define CMT_DISABLE_CENTER N_("Disable center alignment")
 #define CMT_FOLD_TEXTAREA N_("Fold lines in TEXTAREA")
+#define CMT_FOLD_PRE N_("Fold lines in PRE")
 #define CMT_DISP_INS_DEL N_("Display INS, DEL, S and STRIKE element")
 #define CMT_COLOR        N_("Display with color")
 #define CMT_HINTENSITY_COLOR N_("Use high-intensity colors")
@@ -424,6 +425,8 @@ struct param_ptr params1[] = {
      CMT_DISABLE_CENTER, NULL},
     {"fold_textarea", P_CHARINT, PI_ONOFF, (void *)&FoldTextarea,
      CMT_FOLD_TEXTAREA, NULL},
+    {"fold_pre", P_CHARINT, PI_ONOFF, (void *)&FoldPre,
+     CMT_FOLD_PRE, NULL},
     {"display_ins_del", P_INT, PI_SEL_C, (void *)&displayInsDel,
      CMT_DISP_INS_DEL, displayinsdel},
     {"ignore_null_img_alt", P_INT, PI_ONOFF, (void *)&ignore_null_img_alt,
--- a/doc-de/README.func
+++ b/doc-de/README.func
@@ -31,6 +31,7 @@ EXEC_SHELL	Führe Shell-Befehl aus und z
 EXIT		Sofort beenden
 EXTERN		Verwende externen Browser zur Anzeige
 EXTERN_LINK	Verwende externen Browser zur Anzeige des Linkziels
+FOLD_PRE	Erzwinge Umbruch von langen Zeilen in <pre>-Elementen
 FRAME		Wechsle zwischen Kennung und Umsetzung von HTML-Frames
 GOTO		Öffne angegebenes Dokument in neuem Puffer
 GOTO_HOME	Zurück zur Startseite (die Variablen HTTP_HOME oder WWW_HOME spezifiziert wurden)
--- a/doc/README.func
+++ b/doc/README.func
@@ -31,6 +31,7 @@ EXEC_SHELL	Execute shell command and dis
 EXIT		Quit without confirmation
 EXTERN		Display using an external browser
 EXTERN_LINK	Display target using an external browser
+FOLD_PRE	Fold long lines in <pre> elements
 FRAME		Toggle rendering HTML frames
 GOTO		Open specified document in a new buffer
 GOTO_HOME	Return to the homepage (specified HTTP_HOME or WWW_HOME variable)
